File: src/lib/google.ts
export type GoogleUser = {
	sub: string;
	email: string;
	name: string;
	picture?: string;
};

export async function verifyGoogleIdToken(
	idToken: string,
	clientId: string | undefined,
): Promise<GoogleUser | null> {
	try {
		const res = await fetch(
			`https://oauth2.googleapis.com/tokeninfo?id_token=${encodeURIComponent(idToken)}`,
		);
		if (!res.ok) return null;
		const data = (await res.json()) as {
			sub: string;
			email: string;
			name?: string;
			aud?: string;
			email_verified?: string;
		};
		if (clientId && data.aud !== clientId) return null;
		if (!data.email || data.email_verified === "false") return null;
		return {
			sub: data.sub,
			email: data.email,
			name: data.name ?? data.email.split("@")[0],
		};
	} catch {
		return null;
	}
}

export function getGoogleClientId(env: Env): string | undefined {
	return (env as Env & { GOOGLE_CLIENT_ID?: string }).GOOGLE_CLIENT_ID;
}

File: backend/app/services/google.py
import httpx

from app.config import settings


async def verify_google_id_token(id_token: str) -> dict | None:
    if not id_token:
        return None
    try:
        async with httpx.AsyncClient(timeout=15.0) as client:
            res = await client.get(
                "https://oauth2.googleapis.com/tokeninfo",
                params={"id_token": id_token},
            )
        if not res.is_success:
            return None
        data = res.json()
        if settings.google_client_id and data.get("aud") != settings.google_client_id:
            return None
        if not data.get("email") or data.get("email_verified") == "false":
            return None
        return {
            "sub": data["sub"],
            "email": data["email"].lower(),
            "name": data.get("name") or data["email"].split("@")[0],
        }
    except Exception:
        return None
